src/electrum.rs

```diff
@@ -275,6 +275,88 @@ impl Rpc {
         Ok(status)
     }
 
+    fn outpoint_subscribe(&self, (txid, vout): (Txid, u32)) -> Result<Value> {
+        let funding = OutPoint { txid, vout };
+
+        let funding_blockhash = self.tracker.get_blockhash_by_txid(funding.txid);
+        let spending_blockhash = self.tracker.get_blockhash_spending_by_outpoint(funding);
+
+        let mut inputs_funding_confirmed = true;
+
+        let funding_tx = self.daemon.get_transaction(&funding.txid, funding_blockhash);
+
+        if funding_tx.is_err() {
+            return Ok(json!({}));
+        }
+
+        let funding_inputs = &funding_tx.as_ref().unwrap().input;
+        let outpoint_script = &funding_tx.as_ref().unwrap().output.get(vout as usize).unwrap().script_pubkey;
+
+        if funding_blockhash.is_none() && funding_inputs.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()) {
+            inputs_funding_confirmed = false;
+        }
+
+
+        let scripthash = ScriptHash::new(&outpoint_script);
+        let outpoint_scripthash_status = self.new_status(scripthash);
+
+        let funding_height = self.tracker.chain().get_block_height(&funding_blockhash.unwrap()).unwrap();
+
+        if spending_blockhash.is_none() {
+            let txids: Vec<Txid> = outpoint_scripthash_status.unwrap()
+                .get_mempool(&(self.tracker).mempool())
+                .iter()
+                .filter_map(|tx_entry| {
+                    let mempool_tx = self.daemon.get_transaction(&tx_entry.txid, None);
+                    if is_spending(&mempool_tx.unwrap(), funding) {
+                        Some(tx_entry.txid)
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            if txids.len() > 1 {
+                return Ok(panic!("double spend of {}: {:?}", txid, txids));
+            }
+            if !txids.is_empty() {
+                let spending_tx = self.daemon.get_transaction(&txids[0], spending_blockhash);
+                if spending_tx.unwrap().input.iter().any(|txi| self.tracker.get_blockhash_by_txid(txi.previous_output.txid).is_none()) {
+                    if inputs_funding_confirmed {
+                        return Ok(json!({"height": 0, "spender_txhash": txids[0], "spender_height": -1}));
+                    }
+                    return Ok(json!({"height": -1, "spender_txhash": txids[0], "spender_height": -1}));
+                }
+                return Ok(json!({"height": funding_height, "spender_txhash": txids[0], "spender_height": 0}));
+            }
+            return Ok(json!({"height": funding_height}));
+        }
+
+        let spending_height = self.tracker.chain().get_block_height(&spending_blockhash.unwrap()).unwrap();
```

Comment:
Another unidiomatic code. This one is a bit tricky but I think `let mut iter = match spending_blockhash { ... };` Should do the trick. `Either` from crate `either` will need to be used. This can prevent allocation. You can then check for double spends:

```rust
let spender_txid = iter.next().ok_or_else(|| anyhow::anyhow!("WTF spending tx not found but exist ?"));
if let Some(double_spending_txid) = iter.next() {
    // This will probably never happen so allocating is OK.
    let mut txids = vec![spender_txid, double_spending_txid];
    txids.extend(iter);
    bail!("double spend of {}: {:?}", txid, txids);
}
```

Reply:
I think I understand what your suggestion is doing.

Notice that it is totally normal if we don't find any spending txid. The method can be called on a outpoint that is unspend and we have to return a JSON with only the height of the funding transaction. So I guess I have to change the argument in `ok_or_else` method of your suggestion.

Reply:
Ah, right then instead of `ok_or_else` probably `match` and `None` should return the height as success?

Reply:
I end up refactoring most of the code, it should be more idiomatic now ! Feel free to tell me if there is something else ! :)